dask/dataframe/groupby.py

```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+        dataframe and should return a dataframe of intermediate values with the
+        group keys as an index.
+
+    finalizers: a list of result-column transform pairs, that are applied after
+        the ``agg_funcs``. They are used to create final results from
+        intermediate representations.
+    """
+    # generator for consecutive IDs (for intermediate results)
+    ids = it.count()
+    next_id = lambda prefix: '{}-{}'.format(prefix, next(ids))
+
+    simple_impl = {
+        'sum': (M.sum, M.sum),
+        'min': (M.min, M.min),
+        'max': (M.max, M.max),
+        'count': (M.count, M.sum),
+        'size': (M.size, M.sum),
+    }
+
+    known_np_funcs = {np.min: 'min', np.max: 'max'}
+
+    chunks = []
+    aggs = []
+    finalizers = []
+
+    for (result_column, func, input_column) in spec:
+        func = funcname(known_np_funcs.get(func, func))
+
+        if func in simple_impl.keys():
+            impls = _build_agg_args_simple(result_column, func, input_column,
+                                           next_id, simple_impl[func])
+
+        elif func == 'var':
+            raise NotImplementedError()
+
+        elif func == 'std':
+            raise NotImplementedError()
+
+        elif func == 'mean':
+            impls = _build_agg_args_mean(result_column, func, input_column,
+                                         next_id)
+
+        elif func == 'nunique':
+            raise NotImplementedError()
+
+        else:
+            raise ValueError("unknown aggregate {}".format(func))
+
+        chunks.extend(impls.chunk_funcs)
+        aggs.extend(impls.aggregate_funcs)
+        finalizers.extend(impls.finalizers)
+
+    return chunks, aggs, finalizers
+
+
+AggArgs = collections.namedtuple('AggArgs', ['chunk_funcs', 'aggregate_funcs',
+                                             'finalizers'])
+
+
+def _build_agg_args_simple(result_column, func, input_column, next_id, impl_pair):
+    intermediate = next_id(func)
+    chunk_impl, agg_impl = impl_pair
+
+    return AggArgs(
+        chunk_funcs=[partial(_apply_func_to_column, intermediate, input_column,
+                             chunk_impl)],
+        aggregate_funcs=[partial(_apply_func_to_column, intermediate,
+                                 intermediate, agg_impl)],
+        finalizers=[(result_column, operator.itemgetter(intermediate))],
+    )
+
+def _build_agg_args_mean(result_column, func, input_column, next_id):
+    int_sum = next_id('sum')
+    int_count = next_id('count')
+
+    return AggArgs(
+        chunk_funcs=[
+            partial(_apply_func_to_column, int_sum, input_column, M.sum),
+            partial(_apply_func_to_column, int_count, input_column, M.count),
+        ],
+        aggregate_funcs=[
+            partial(_apply_func_to_column, int_sum, int_sum, M.sum),
+            partial(_apply_func_to_column, int_count, int_count, M.sum)
+        ],
+        finalizers=[(result_column, partial(_finalize_mean, int_sum, int_count))],
```

Comment:
Thanks for the explanation. To be hones, I was wondering when reading the dask code base about this design choice, makes perfect sense. 

I'll put the arguments directly into the graph. My proposal would be to reverse the arguments to `_apply_func_to_column`, convert all partial arguments to keyword arguments, and to replace the partials with pairs of function and kwargs.